relay-server/src/actors/processor/dynamic_sampling.rs

```diff
@@ -96,20 +96,21 @@ pub fn run(state: &mut ProcessEnvelopeState, config: &Config) {
 }
 
 /// Apply the dynamic sampling decision from `compute_sampling_decision`.
-pub fn sample_envelope(state: &mut ProcessEnvelopeState) -> Result<(), ProcessingError> {
+pub fn sample_envelope_items(state: &mut ProcessEnvelopeState) {
     if let SamplingResult::Match(sampling_match) = std::mem::take(&mut state.sampling_result) {
         // We assume that sampling is only supposed to work on transactions.
         if state.event_type() == Some(EventType::Transaction) && sampling_match.should_drop() {
             let matched_rules = sampling_match.into_matched_rules();
 
+            let outcome = Outcome::FilteredSampling(matched_rules.clone());
             state
                 .managed_envelope
-                .reject(Outcome::FilteredSampling(matched_rules.clone()));
+                .retain_items(|_| utils::ItemAction::Drop(outcome.clone()));
 
-            return Err(ProcessingError::Sampled(matched_rules));
+            // The event is no longer in the envelope, so we need to handle it separately:
+            state.reject_event(outcome);
```

Comment:
I don't quite understand this. We drop the items in the managed envelope through retain_items which seems to already log outcomes through the managed envelope, why is this still needed?

Reply:
The event at this point is not an envelope item, but a separate field on the state (see `extract_event`).

> And why the event_category conditional check in the implementation?

We need an event category to log an outcome for it. When `event_category` is `None`, it means there is no event to remove.